### PR TITLE
Issue 1267 - Remove .success_type and .failure_type

### LINKED ISSUE #1267: Remove `.success_type` and `.failure_type`
Since `Success` / `Failure` and `Some` / `Nothing` are real types, we can remove `.success_type` and `.failure_type` hacks.

It is a breaking change, but it is a simple one to fix.
And it is a better API.

### PR DESCRIPTION
**Pull request by ducdetronquito**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues
- Closes #1267

The process to start contributing is very clear: it's nice when it's easy to get started :)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,14 @@ incremental in minor, bugfixes only are patches.
 See [0Ver](https://0ver.org/).
 
 
+## 0.23.0
+
+### Misc
+
+- *Breaking*: Remove `Result.success_type` and `Result.failure_type`
+- *Breaking*: Remove `Maybe.success_type` and `Maybe.failure_type`
+- *Breaking*: Remove `IOResult.success_type` and `Maybe.failure_type`
```
Comment (thepabloaguilar) [inline CHANGELOG.md:15]:
```suggestion
- *Breaking*: Remove `IOResult.success_type` and `IOResult.failure_type`
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,14 @@ incremental in minor, bugfixes only are patches.
 See [0Ver](https://0ver.org/).
 
 
+## 0.23.0
```
Comment (thepabloaguilar) [inline CHANGELOG.md:9]:
```suggestion
## 0.23.0 WIP
```

File: tests/test_maybe/test_maybe_base.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
-from returns.maybe import Maybe
+from returns.maybe import Some, _Nothing
```
Comment (thepabloaguilar) [inline tests/test_maybe/test_maybe_base.py:1]: This test can be deleted, the only purpose was to ensure we put a type inside `Maybe.success_type` and `Maybe.failure_type`

File: tests/test_result/test_result_base.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
-from returns.result import Result
+from returns.result import Failure, Success
```
Comment (thepabloaguilar) [inline tests/test_result/test_result_base.py:1]: This test can be deleted, the only purpose was to ensure we put a type inside `Result.success_type` and `Result.failure_type`

File: typesafety/test_io/test_ioresult_container/test_ioresult_helpers.yml [hunk 5]
```diff
@@ -32,14 +32,14 @@
...
 
 
 - case: ioresult_failure_type
   disable_cache: false
   main: |
-    from returns.io import IOResult
+    from returns.io import IOFailure
 
-    reveal_type(IOResult.failure_type)  # N: Revealed type is "Type[returns.io.IOFailure[Any]]"
+    reveal_type(IOFailure)  # N: Revealed type is "Type[returns.io.IOFailure[Any]]"
```
Comment (thepabloaguilar) [inline typesafety/test_io/test_ioresult_container/test_ioresult_helpers.yml:45]: These tests can be deleted as we don't have `.{success | failure}_type` anymore

File: typesafety/test_maybe/test_maybe_type_cast.yml [hunk 6]
```diff
@@ -36,15 +36,15 @@
...
 
 
 - case: maybe_result_failure_type
   disable_cache: false
   main: |
-    from returns.maybe import Maybe
+    from returns.maybe import _Nothing
 
-    reveal_type(Maybe.failure_type)  # N: Revealed type is "Type[returns.maybe._Nothing]"
+    reveal_type(_Nothing)  # N: Revealed type is "Type[returns.maybe._Nothing]"
```
Comment (thepabloaguilar) [inline typesafety/test_maybe/test_maybe_type_cast.yml:49]: These tests can be deleted as we don't have `.{success | failure}_type` anymore

File: typesafety/test_result/test_result_type_cast.yml [hunk 7]
```diff
@@ -128,17 +128,17 @@
...
 
 
 - case: result_failure_type
   disable_cache: false
   main: |
-    from returns.result import Result
+    from returns.result import Failure
 
-    reveal_type(Result.failure_type)  # N: Revealed type is "Type[returns.result.Failure[Any]]"
+    reveal_type(Failure)  # N: Revealed type is "Type[returns.result.Failure[Any]]"
```
Comment (thepabloaguilar) [inline typesafety/test_result/test_result_type_cast.yml:141]: These tests can be deleted as we don't have `.{success | failure}_type` anymore

**Review by thepabloaguilar — CHANGES_REQUESTED** [review]

**Review by thepabloaguilar — COMMENTED** [review]
Thanks for your contribution @ducdetronquito!

**Comment by ducdetronquito** [comment]
Hi @thepabloaguilar ,

Thank you for your review !

I made the requested changes in a dedicated commit to ease your review but if you prefer I can edit the original commits.

Have a nice day !

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,14 @@ incremental in minor, bugfixes only are patches.
...
+
+### Misc
+
+- *Breaking*: Remove `Result.success_type` and `Result.failure_type`
```
Comment (sobolevn) [inline CHANGELOG.md:13]: Let's list all removed items in a single point

Comment (ducdetronquito) [reply CHANGELOG.md]: I made the change in a new commit

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,12 @@ incremental in minor, bugfixes only are patches.
 See [0Ver](https://0ver.org/).
 
 
+## 0.23.0 WIP
```
Comment (sobolevn) [inline CHANGELOG.md:9]: Let's make it `1.0.0 WIP` :)

Comment (thepabloaguilar) [reply CHANGELOG.md:9]: Oh, 1.0.0 already? 👀

Comment (ducdetronquito) [reply CHANGELOG.md]: Done deal :D

**Review by sobolevn — APPROVED** [review]
Thank you!

**Pull request merged**